pennylane/math/quantum.py

```diff
@@ -606,3 +606,158 @@ def _compute_mutual_info(
+    systems with the same size. For two pure states, the fidelity corresponds to the squared overlap. For a pure state
+    and a mixed state, it corresponds to the squared expectation of the mixed state in the pure state. Finally for two
+    mixed states, it is defined by the last formula:
+
+    .. math::
+        F( \ket{\psi} , \ket{\phi}) = \left|\bra{\psi}\ket{\phi}\right|^2
+
+        F( \ket{\psi} , \sigma ) = \left|\bra{\psi} \sigma \ket{\psi}\right|^2
+
+        F( \rho , \sigma ) = -\text{Tr}( \sqrt{\sqrt{\rho} \sigma \sqrt{\rho}})^2
+
+    .. warning::
+        The second state is coerced to the type and dtype of the first state. The fidelity is returned in the type
+        of the interface of the first state.
+
+    Args:
+        state0 (tensor_like): 1D state vector or 2D density matrix
+        state1 (tensor_like): 1D state vector or 2D density matrix
+        check_state (bool): If True, the function will check the state validity (shape, norm) or (shape, trace,
+            semi positive definitiveness).
+        c_dtype (str): Complex floating point precision type.
+
+    Returns:
+        float: Fidelity between the two quantum states.
+
+    **Example**
+
+        >>> state0 = [0.98753537-0.14925137j, 0.00746879-0.04941796j]
+        >>> state1 = [0.99500417+0.j, 0.09983342+0.j]
+        >>> qml.math.fidelity(state0, state1)
+        0.9905158135644924
+
+        >>> state0 = [0, 1]
+        >>> state1 = [[0, 0], [0, 1]]
+        >>> qml.math.fidelity(state0, state1)
+        1.0
+
+        >>> state0 = [[1, 0], [0, 0]]
+        >>> state1 = [[0, 0], [0, 1]]
+        >>> qml.math.fidelity(state0, state1)
+        0.0
+
+    """
+    # Cast as a c_dtype array
+    state0 = cast(state0, dtype=c_dtype)
+    len_state0 = state0.shape[0]
+
+    # Cannot be cast_like if jit
+    if not is_abstract(state0):
+        state1 = cast_like(state1, state0)
+
+    len_state1 = state1.shape[0]
+
+    if check_state:
+        if state0.shape == (len_state0,):
+            _check_state_vector(state0)
+        else:
+            _check_density_matrix(state0)
+
+        if state1.shape == (len_state1,):
+            _check_state_vector(state1)
+        else:
+            _check_density_matrix(state1)
+
+    # Get dimension of the quantum system and reshape
+    num_indices0 = int(np.log2(len_state0))
+    num_indices1 = int(np.log2(len_state1))
+
+    if num_indices0 != num_indices1:
+        raise qml.QuantumFunctionError("The two states must have the same number of wires.")
+
+    # Two pure states, squared overlap
+    if state1.shape == (len_state1,) and state0.shape == (len_state0,):
+        overlap = np.tensordot(state0, np.transpose(np.conj(state1)), axes=1)
+        overlap = np.real(overlap) ** 2
+        return overlap
+    # First state mixed, second state pure
+    if state1.shape == (len_state1,) and state0.shape != (len_state0,):
+        overlap = np.tensordot(state0, np.transpose(np.conj(state1)), axes=1)

```

Comment:
Again purely subjective, but if you use `tensordot(x, y, axes=1)` you might as well just `x @ y`, which I find more readable.

Especially in this case because then it just becomes `real(conj(transpose(x)) @ sigma @ x)` in one line.

Reply:
It does not work with TF unfortunately :(